File: csharp/200-number-of-islands.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        var rows = grid.Length;
        var cols = grid[0].Length;
        var visited = new HashSet<(int, int)>();
        var islands = 0;

        void dfs(int row, int col, char[][] grid) {
            if (row < 0 || row >= rows || col < 0 || col >= cols) {
                return;
            }
            if (visited.Contains((row, col)) || grid[row][col] == '0') {
                return;
            }

            visited.Add((row, col));

            dfs(row + 1, col, grid);
            dfs(row - 1, col, grid);
            dfs(row, col + 1, grid);
            dfs(row, col - 1, grid);
        }

        for(var row = 0; row < rows; row++) {
            for(var col = 0; col < cols; col++) {
                if (!visited.Contains((row, col)) && grid[row][col] == '1') {
                    islands++;
                    dfs(row, col, grid);
                }
            }
        }
        return islands;
    }
}

File: python/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        rows, cols = len(grid), len(grid[0])
        visit = set()
        islands = 0
        
        def bfs(row: int, col: int) -> None:
            q = deque()
            visit.add((row, col))
            q.append((row, col))

            while q:
                r, c = q.popleft()
                directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

                for dr, dc in directions:
                    if ((r + dr) in range(rows) and
                        (c + dc) in range(cols) and
                        grid[r + dr][c + dc] == '1' and
                        (r + dr, c + dc) not in visit):
                        q.append((r + dr, c + dc))
                        visit.add((r + dr, c + dc))
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == '1' and (row, col) not in visit:
                    bfs(row, col)
                    islands += 1
        return islands
        